File: 백준/Gold/15686. 치킨 배달/치킨 배달.py
import sys

n,m = map(int, input().split())
g = [list(map(int, input().split())) for _ in range(n)]
# 0 빈칸 1 집 2 치킨집

#먼저 치킨집 m개 선택
home = []
pizza = []
for i in range(n):
    for j in range(n):
        if g[i][j] == 1:
            home.append((i,j))
        if g[i][j] == 2:
            pizza.append((i,j))
k = len(pizza) #피자집 개수
choose = [] #선택한 피자집 인덱스

def min_length(choose):
    length = 0
    for i in range(len(home)): #집을 기준으로
        dis = int(1e9)
        x1,y1 = home[i][0], home[i][1]
        for idx in  choose: #피자집 인덱스 하나씩 꺼내서
            x2,y2 = pizza[idx][0], pizza[idx][1]
            distance = abs(x1-x2) + abs(y1-y2)
            dis = min(dis, distance)
            #i번째 집의 치킨거리 dis로 결정
        length += dis
    return length #선택된 치킨집에 대한 도시의 최소거리 리턴
res = int(1e9)
def DFS(L, start): #
    global res
    if L == m: #피자집 모두 선택했다면 이제 도시의 피자거리 구해야지
        length = min_length(choose)
        res = min(res, length)
    else: #피자집부터 선택
        for i in range(start, k):
            choose.append(i)
            DFS(L+1, i+1) #현재꺼 선택했으면 다음에는 선택 안해야 하므로
            choose.pop()


DFS(0,0)
print(res)

File: 백준/Gold/15686. 치킨 배달/치킨 배달.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;


class Point{
    int x, y;

    Point(int x, int y) {
        this.x = x;
        this.y = y;
    }
}
public class Main {
    // 0 빈칸 1 집 2 피자집
    // 피자배달거리 : |x1-y2| + |y1-y2|
    // 피자집 중 m개만 선택.
    static int n,m;
    static int minDis = Integer.MAX_VALUE;
    static int[][] g;
    static List<Point> home, chicken;  // 집, 치킨집 저장
    static int[] res;  // 선택한 치킨집들 인덱스
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        g = new int[n+1][n+1];
        res = new int[m];  // 뽑은 피자집을 저장할 배열
        home = new ArrayList<>();
        chicken = new ArrayList<>();
        for (int i = 1; i <= n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 1; j <= n; j++) {
                g[i][j] = Integer.parseInt(st.nextToken());
                if(g[i][j] == 1) home.add(new Point(i, j));  // 집
                else if(g[i][j] == 2) chicken.add(new Point(i, j));  // 피자집
            }
        }
        // 세팅 끝.
        DFS(0, 0);
        System.out.println(minDis);
    }

    public static void DFS(int start, int L) {
        if (L == m) { // m개 선택. 종료 조건
            // 이제 선택된 m개에서의 최단거리 확인.
            int sum = calDis();
            if (sum < minDis) {
                minDis = sum;
            }
        } else {
            for (int i = start; i < chicken.size(); i++) {
                res[L] = i;
                DFS(i + 1, L + 1);
            }
        }
    }

    public static int calDis() {
        int sum = 0;
        for (int i = 0; i < home.size(); i++) {  // 집 기준으로 계산해야함.
            int dis = Integer.MAX_VALUE;
            for (int j = 0; j < res.length; j++) {
                dis = Math.min(dis, Math.abs(home.get(i).x - chicken.get(res[j]).x) + Math.abs(home.get(i).y - chicken.get(res[j]).y));
            }
            sum += dis;
        }
        return sum;
    }
}